### PR TITLE
Ledger export and migrate cmd

### PR DESCRIPTION
**Pull request by ramtinms**
This PR
- replaces the old extract state cmd with the new ledger functionality (construct a new trie instead of pruning), this enables pathfinder version migrations and make other migrations less error-prone 
- introduced ledger reporter interface, a way to produce a summary of the state of the ledger (which will be used in the upcoming PRs). This will be super helpful to check the health of the state after migration and before spork
- makes block_id an optional flag for this cmd, enabling the operation by state commitment as well.

### CODE REVIEW

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go [hunk 1]
```diff
@@ -28,134 +20,20 @@ func getStateCommitment(commits storage.Commits, blockHash flow.Identifier) (flo
...
-
-	err = wal.StoreCheckpoint(flattenForest, checkpointWriter)
-	if err != nil {
-		return fmt.Errorf("cannot store checkpoint: %w", err)
-	}
-
-	duration = time.Since(startTime)
-	log.Info().Float64("total_time_s", duration.Seconds()).Msg("finished writing checkpoint")
-
+	log.Info().Msg("New state commitment for the exported state is :" + newState.String())
```
Comment (janezpodhostnik) [inline cmd/util/cmd/execution-state-extract/execution_state_extract.go:37]: should we add a base64 encoded state commitment as well? just for convenience, as that is what is used in the jsons...

**Review by janezpodhostnik — APPROVED** [review]

File: cmd/util/cmd/execution-state-extract/cmd.go [hunk 2]
```diff
@@ -34,37 +35,49 @@ func init() {
...
-	stateCommitment, err := getStateCommitment(commits, blockID)
-	if err != nil {
-		log.Fatal().Err(err).Msg("cannot get state commitment for block")
+		stateCommitment, err = getStateCommitment(commits, blockID)
+		if err != nil {
+			log.Fatal().Err(err).Msg("cannot get state commitment for block")
+		}
+	}
+
+	if len(flagStateCommitment) > 0 {
```
Comment (tarakby) [inline cmd/util/cmd/execution-state-extract/cmd.go:69]: I see that if both flags are entered, priority is given to `flagBlockHash`. Shouldn't we not allow having both flags to make it clearer?

File: ledger/ledger.go [hunk 3]
```diff
@@ -274,4 +274,9 @@ func (v Value) MarshalJSON() ([]byte, error) {
 }
 
 // Migration defines how to convert the given slice of input payloads into an slice of output payloads
-type Migration func(payload []Payload) ([]Payload, error)
+type Migration func(payloads []Payload) ([]Payload, error)
+
+// Reporter accpets an slice ledger payloads and report the state of the ledger
```
Comment (tarakby) [inline ledger/ledger.go:279]:
```suggestion
// Reporter accepts slice ledger payloads and reports the state of the ledger
```

**Review by tarakby — APPROVED** [review]

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go (hunk 1 above)
Comment (ramtinms) [reply cmd/util/cmd/execution-state-extract/execution_state_extract.go:37]: do we use base64 encoding of state commitments anywhere? String() will return the hex-encoded

File: cmd/util/cmd/execution-state-extract/cmd.go (hunk 2 above)
Comment (ramtinms) [reply cmd/util/cmd/execution-state-extract/cmd.go:69]: Yup, I'll add the check and update the description.

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go (hunk 1 above)
Comment (janezpodhostnik) [reply cmd/util/cmd/execution-state-extract/execution_state_extract.go:37]: root-block-seal.json and root-execution-result.json have a base64 encoded state commitment.

Comment (ramtinms) [reply cmd/util/cmd/execution-state-extract/execution_state_extract.go]: right right, I forgot that again

Comment (ramtinms) [reply cmd/util/cmd/execution-state-extract/execution_state_extract.go]: I added at de2250d

File: cmd/util/cmd/execution-state-extract/cmd.go (hunk 2 above)
Comment (ramtinms) [reply cmd/util/cmd/execution-state-extract/cmd.go:74]: I added at c32ce98

**Pull request merged**